Added Docker File

Abhinavks1405 opened the pull request.

SaashaJoshi commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM python:3.9-slim
```
Can this be edited to support python versions from 3.8 to 3.12?

SaashaJoshi commented on Dockerfile:
```diff
@@ -0,0 +1,7 @@
+FROM python:3.9-slim
+WORKDIR /app
+COPY requirements.txt /app/requirements.txt
+RUN python -m venv venv
```
I am unsure if `venv` needs to be initiated when creating a docker image. Do you have a particular reason behind doing so?

Abhinavks1405 replied:
Having virtual env in docker ensures that there is consistency accors different environment and not have a package conflict

SaashaJoshi replied:
But Docker itself creates separate containers for this purpose.

Abhinavks1405 replied:
Understood , i thought we are working with multiple python version in the same docker , which is not he case

SaashaJoshi reviewed (changes requested):
Hi @Abhinavks1405, thanks for the input. I just left a few reviews and the rest LGTM.

SaashaJoshi commented:
Also, do have a look at the errors while lining the file. Let me know if you will need any help there!

Abhinavks1405 commented:
@SaashaJo

Abhinavks1405 commented:
@SaashaJoshi 
I have commited the desired changes.
1. added support for python version 3.8 to 3.12.
2. updated pylint-> v3.1.0 to remove the error occurred while linting with python 3.12 .
3. added build.yml to have github action workflows .

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,21 @@
+name: Build and Test
+
+on: [push, pull_request]
```
I would prefer if Docker images were made only on a workflow dispatch.

```suggestion
on: 
    workflow_dispatch:
```

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,21 @@
+name: Build and Test
```
Could you rename the file to Build Docker? Also `build_docker.yml"

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,21 @@
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
+
+    steps:
+    - uses: actions/checkout@v2
```
You can now use `checkout@v3`

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,21 @@
+
+    - name: Build Docker image
+      run: |
+        docker build --build-arg PYTHON_VERSION=${{ matrix.python-version }} -t your-app-name:${{ matrix.python-version }} .
```
I think `your-app-name` is a placeholder that should probably be replaced with `piqture`

SaashaJoshi replied:
Also, `piqture` app versions are different from python versions. You will have to add another strategy matrix that takes into account the app version.

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,21 @@
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
```
```suggestion
        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
        app_veriosn: ["v0.1.1"]
```

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,21 @@
+
+    - name: Run Docker container
+      run: |
+        docker run your-app-name:${{ matrix.python-version }}
```
The `your-app` placeholder needs to be changed here as well.

SaashaJoshi commented on .github/workflows/unit_test.yml:
```diff
@@ -40,3 +40,4 @@ jobs:
       run: coveralls
       env:
         COVERALLS_REPO_TOKEN: ${{ secrets.COVERALLS_REPO_TOKEN }}
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }} #Github Tokens
```
This is not the correct place to have a GitHub token. You will have to provide me with the token and I will add it as a secret on GitHub.

SaashaJoshi replied:
Which token are we trying to implement here?

Abhinavks1405 replied:
My bad , I didn't noticed that github tokens got exposed.

Abhinavks1405 replied:
the github token is automatically set and it is required because without it the upload coverage to coverall is failing

SaashaJoshi commented on .pre-commit-config.yaml:
```diff
@@ -31,7 +31,7 @@ repos:
 
 - repo: https://github.com/PyCQA/pylint
-  rev: v2.17.0
+  rev: v3.1.0
```
This is an unrelated change for the PR. You will have to start another PR to make this change.

SaashaJoshi commented on Dockerfile:
```diff
@@ -0,0 +1,12 @@
+ARG PYTHON_VERSION=3.12
```
Can we have a default value of 3.9

Abhinavks1405 commented:
i have checked and done all the possible changes , the only issue currently facing and previously mentioned is of github tokens .

Abhinavks1405 commented:
@SaashaJoshi  , waiting for you review on pr .

SaashaJoshi commented:
Hey, can you revert all the changes made to `unit-test.yml` and `.pre-commit.yml`? These files weren't meant to be changed.

Abhinavks1405 commented:
@SaashaJoshi reverted changes in unit_test.yml and .pre-commit-config.yaml

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,49 @@
+    strategy:
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
+    outputs:
```
We do not need outputs from our yml file. Can we safely remove this?

SaashaJoshi commented:
There is only one thing that I notice here. We are building Docker images twice for each version and each app version. However, what we wish to do is create Docker images for `piQturev0.1.1` with each python version. Let me suggest some changes.

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,49 @@
+
+    - name: Build Docker image for Python versions
+      run: |
+        docker build --build-arg PYTHON_VERSION=${{ matrix.python-version }} \
+                     -t piqture:${{ matrix.python-version }} .
```
```suggestion
        docker build --build-arg PYTHON_VERSION=${{ matrix.python-version }} \
                     --build-arg APP_VERSION=${{matrix.app-version}} \
                     -t piqture:${{ matrix.app-version }}-${{ matrix.python-version }} .
```

SaashaJoshi replied:
With these changes, we will not require the `build-app` yml workflow.

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,49 @@
+    - name: Set up Docker
+      uses: docker/setup-buildx-action@v2
+
+    - name: Build Docker image for Python versions
```
```suggestion
    - name: Build Docker image for App and Python versions
```

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,49 @@
+        docker build --build-arg PYTHON_VERSION=${{ matrix.python-version }} \
+                     -t piqture:${{ matrix.python-version }} .
+
+  build-app:
+    runs-on: ubuntu-latest
+    needs: build
+    strategy:
+      matrix:
+        app-version: ['v0.1.1']
+      max-parallel: 1
+    steps:
+    - uses: actions/checkout@v3
+
+    - name: Set up Docker
+      uses: docker/setup-buildx-action@v2
+
+    - name: Pull base Docker image from previous build
+      run: |
+        docker pull user/piqture:${{needs.build.outputs.python-version}}
+
+    - name: Build Docker image for App versions
+      run: |
+        docker build --build-arg APP_VERSION=${{ matrix.app-version }} \
+                     -t piqture-app:${{ matrix.app-version }} .
+
+    - name: Run Docker container
+      run: |
+        docker run piqture-app:${{ matrix.app-version }}
```
Any of this will not be needed with the above suggested changes.

SaashaJoshi commented on .github/workflows/build.yml:
```diff
@@ -0,0 +1,49 @@
+      matrix:
+        python-version: ['3.8', '3.9', '3.10', '3.11', '3.12']
+    outputs:
+      python-version: $({matrix.python-version})
```
We do not need this variable. Instead define `app-version` here.
```suggestion
      app-version: ["0.1.1"]
```

SaashaJoshi reviewed (commented):
Suggested changes to reduce the Docker workflow to build app and python versions of the app once only. Let me know what you think!

Abhinavks1405 commented:
> There is only one thing that I notice here. We are building Docker images twice for each version and each app version. However, what we wish to do is create Docker images for `piQturev0.1.1` with each python version. Let me suggest some changes.

the point which I processed was mainly scalability as having 2 matrices to handle python version and app version so that we don't have to continuously change logic each time , mainly having 2 docker build ensures that I can have multiple app version running on same python version image reducing build time . i maily proceeded with this logic but want to know your thoughts as they are helpful .

SaashaJoshi commented:
I agree with your point, But doesn't it mean we will have, let's say 5 Docker versions per app version? This is also what I am trying to achieve with a single Docker build. For example, the build will produce `piQture-1.1-3.8` then `piQture-1.1-3.9`, and so on.

Abhinavks1405 commented:
> I agree with your point, But doesn't it mean we will have, let's say 5 Docker versions per app version? This is also what I am trying to achieve with a single Docker build. For example, the build will produce `piQture-1.1-3.8` then `piQture-1.1-3.9`, and so on.

Ok I understand your point , I can try to do this with tags to bring it to a single docker build otherwise the app version can be hardcoded .

SaashaJoshi commented:
Hi @Abhinavks1405, any updates? I will try to get this PR merged by today. You can then open a PR to make corrective measures after the hacktoberfest marked PR is merged.

Abhinavks1405 commented:
> Hi @Abhinavks1405, any updates? I will try to get this PR merged by today. You can then open a PR to take corrective measures after the hacktoberfest marked PR is merged.

yes yes the changes are made , let me push it

Abhinavks1405 commented:
@SaashaJoshi, can I raise 2 pr after this, one related to the updated Pylint version for supporting python 3.12 and other one for docker file optimizations ?

SaashaJoshi commented:
Sure, feel free to create as many PRs. But just make sure that each PR serves a single purpose.
